### DISCUSSION TITLE
Tilde expansion?

### DISCUSSION THREAD

**Discussion by BurtHarris** [PowerShell/PowerShell#15263] [Q&A] [answered]
Where does bash-style tilde expansion stand in powershell?

**Comment by rkeithhill** [comment, marked as answer]
For built-in commands, PowerShell already supports this.  For native applications, there's an an experimental feature to support this.  I forget which version it shipped in but it's in the latest stable `7.1.3`.  Enable the feature like so:
```powershell
Enable-ExperimentalFeature PSNativePSPathResolution
```
You'll need to restart PowerShell for this change to take affect.  After enabling this feature, you can then do this and it will work:
```powershell
notepad ~\Documents\PowerShell\profile.ps1
```

**Reply by BurtHarris** [reply]
Hmm, Thanks.   I tried this under PowerShell 7.1.3:
``` PowerShell
PS C:\Windows\System32> echo ~
~

PS C:\Windows\System32> bash
burt@Gladiator:/mnt/c/Windows/System32$ echo ~
/home/burt

```  

Did I misunderstand what you meant by built-in commands?

**Reply by rkeithhill** [reply]
`echo` is an alias for `Write-Output` which outputs the string provided to it.  This particular command does not "expand" the ~.  However, this built-in command does expand `~` - `Get-Content ~\Documents\PowerShell\profile.ps1`.  For the built-in commands, those that take a file system path of some sort should work with `~`.

**Reply by rkeithhill** [reply]
The `PSNativePSPathResolution` feature solves a specific problem by expanding `~` in a path when invoking a native application.  Try the example `notepad ~\Documents\PowerShell\profile.ps1` before and after enabling this experimental feature.

**Reply by BurtHarris** [reply]
Ahh, so for cmdlets it depends on the type of parameter?   I guess that covers it.

**Reply by rkeithhill** [reply]
Yup.

**Reply by rkeithhill** [reply]
To be clear, this isn't exactly the same as "globbing" as you've seen with `echo ~`.  But it does solve the most common issue you run into in PS when invoking native apps e.g. `vim ~\.gitconfig`,  `notepad ~\.npmrc`, etc.

**Reply by ninmonkey** [reply]
For commands that don't support globbing, you can pipe from `gi`, `gci`, or `resolve-path`
```ps1
# echo all log files
Get-Item "~\**\*.log" | Get-Content
# or
Resolve-Path "~\**\*.log" | Get-Content
```
That works with `PSNativePSPathResolution = $false`

**Reply by GMZwinge** [reply]
@ninmonkey If the path does not exist, `Get-Item` and `Resolve-Path` fail. In this case `$ExecutionContext.SessionState.Path.GetUnresolvedProviderPathFromPSPath('~\<nonExistentPath>')` seems to work (from [x0n response to Powershell: resolve path that might not exist?](https://stackoverflow.com/questions/3038337/powershell-resolve-path-that-might-not-exist)). But is there something else more simple?

**Comment by alihaiderismail** [comment]
Note that `PSNativePSPathResolution` was removed in Powershell 7.3. The new way to achieve this is `PSNativeWindowsTildeExpansion` which is enabled PowerShell 7.5-preview.2. See [Using Experimental Features in PowerShell](https://learn.microsoft.com/en-us/powershell/scripting/learn/experimental-features?view=powershell-7.4#available-features).

**Comment by roke-julian-lockhart** [comment]
> https://github.com/orgs/PowerShell/discussions/15263#discussion-3326069

@BurtHarris, why not merely use `$HOME`? POSIX or the HFS (I forget which) mandate that `$Env:HOME` exist, and PowerShell creates a `$HOME` variable which points to `$Env:HOME` when `$IsLinux` = `$True`, or `Environment.GetFolderPath(Environment.SpecialFolder.UserProfile)` when `$IsWindows` = `$True`. `~` is a `bash`-ism — not even POSIX-compliant — so I fail to see why supporting that would be desirable.
